panic! doesn't print to console

### Summary

On current `main`, `panic` does not print the message to console.

E.g. exchanging the `info` to `panic` in `examples/hello-world` gives me the following output with "Hello World!" missing.

```bash
panic: panicked at examples/hello-world/src/main.rs:8:5

Frame 0: syscall_readonly @ 0x00001a2a inline
       /home/nils/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.18/src/sys/arm_compat/syscall/arm.rs:88:9
Frame 1: sys_exit_extended @ 0x0000000000001a28 inline
       /home/nils/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.18/src/sys/arm_compat/mod.rs:195:9
Frame 2: exit @ 0x0000000000001a1e inline
       /home/nils/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.18/src/sys/arm_compat/mod.rs:164:5
Frame 3: exit @ 0x0000000000001a1e
       /home/nils/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/semihosting-0.1.18/src/process.rs:57:5
Frame 4: init @ 0x00000224
       /home/nils/Programming/ariel-os/src/ariel-os-embassy/src/lib.rs:130:1
Error: Semihosting indicated exit with reason: 0x20026, exit_status: 1
```

--- a/src/ariel-os-debug/src/lib.rs
+++ b/src/ariel-os-debug/src/lib.rs
@@ -52,6 +52,25 @@ pub fn exit(code: ExitCode) {
     }
 }
 
+/// Prints the panic on the debug output in a consistent manner across loggers.
+#[doc(hidden)]
+pub fn print_panic(info: &core::panic::PanicInfo) {
+    // `location()`'s documentation currently states that it always returns `Some(_)`.
+    // It is unclear what the panic formatting would be otherwise, because the std does not
+    // currently handle the case where the location cannot be obtained.
+    let location = info.location().unwrap();
+    let message = info.message();
+
+    // `PanicMessage` does not currently implement `defmt::Format`.
+    // We *need* to use the `Display` implementation and cannot use `PanicMessage::as_str()` as
+    // that would not work for dynamically formatted messages.
+    #[cfg(feature = "defmt")]
+    let message = ariel_os_debug_log::defmt::Display2Format(&message);
+
+    // Mimics the `Display` implementation of `core::panic::PanicInfo`.
+    println!("panicked at {}:\n{}", location, message);
+}
+
 #[cfg(all(feature = "debug-console", feature = "rtt-target"))]
 mod backend {
     #[cfg(not(feature = "defmt"))]

--- a/src/ariel-os-rt/src/lib.rs
+++ b/src/ariel-os-rt/src/lib.rs
@@ -125,7 +125,7 @@ mod isr_stack {
 #[panic_handler]
 fn panic(_info: &core::panic::PanicInfo) -> ! {
     #[cfg(feature = "panic-printing")]
-    ariel_os_debug::println!("panic: {}\n", _info);
+    ariel_os_debug::print_panic(_info);
 
     ariel_os_debug::exit(ariel_os_debug::ExitCode::FAILURE);
 

